### PR TITLE
Fix UnboundLocalError in calculate_consensus with zero counts

### LINKED ISSUE #5000: UnboundLocalError: local variable 'consensus_letter' referenced before assignment
Hi Biopython team,

I encountered an error when using the `calculate_consensus(identity=...)` method from the `Bio.motifs.matrix.Counts` class. Here is the traceback:

```
  File "align.py", line 146, in main
    consensus = counts.calculate_consensus(identity=0.4)
  File "/home/dslos/.virtualenvs/metabar/lib/python3.10/site-packages/Bio/motifs/matrix.py", line 280, in calculate_consensus
    consensus_letter = consensus_letter.lower()
UnboundLocalError: local variable 'consensus_letter' referenced before assignment
```

**How to reproduce:**
- Create a motif and call `counts.calculate_consensus(identity=0.4)` on it.

**Expected behavior:**  
The method should either return a consensus sequence or raise a meaningful error, but not an `UnboundLocalError`.

**Environment:**  
- Biopython version: 1.85
- Python version: 3.10
- Platform: Linux

**Suggestion:**  
It looks like `consensus_letter` is not always assigned before being used. Please check the logic in `calculate_consensus` to ensure `consensus_letter` is always defined before calling `.lower()` on it.

Thank you!

### PR DESCRIPTION
**Pull request by ernestprovo23**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added tests for the changes I have made.

## Summary

Fix `UnboundLocalError: local variable 'consensus_letter' referenced before assignment` in `calculate_consensus()` when all counts are zero at a position.

Fixes #5000

## Changes

Two-line fix in `Bio/motifs/matrix.py`:
1. Initialize `consensus_letter = undefined` before the inner loop
2. Add `total == 0` guard to the identity threshold check

When a motif is initialized with wrong-case alphabet (e.g. lowercase sequences with uppercase alphabet), all counts are zero. Previously this caused `consensus_letter` to never be assigned, then `.lower()` was called on it. Now it correctly returns the undefined character (`N` for nucleotides, `X` for amino acids).

## Testing

Added 3 regression tests in `Tests/test_motifs.py::TestCalculateConsensus`:
- `test_zero_counts_returns_undefined` - reproduces the exact bug from #5000
- `test_zero_counts_with_identity` - zero counts with identity parameter
- `test_normal_consensus_unchanged` - verifies existing behavior unaffected

```
cd Tests && python -m pytest test_motifs.py::TestCalculateConsensus -v  # 3 passed
```

### CODE REVIEW

**Comment by ernestprovo23** [comment]
credit where its due — @michaelfm1211 spotted the unguarded code path back in july, and @zreitz nailed the reproducer with the case-mismatch alphabet. my fix is basically what they diagnosed.

also worth noting @mdehoon's point that under normal usage `self[letter][i]` should be positive for at least one letter. thats correct for properly constructed motifs. the zero-count case only happens when the alphabet doesnt match the sequences (e.g. uppercase alphabet w/ lowercase seqs), so every count stays at 0. the `total == 0` guard specifically targets that edge case w/o changing behavior for normal inputs.

**Comment by ernestprovo23** [comment]
hey @peterjc — friendly ping on this one. fix is small (3 lines + test) and CI's green. lmk if anything's needed on my end to move it along, happy to rebase or adjust.

**Comment by mdehoon** [comment]
In your test:
```python
        alignment = Alignment(
            [Seq("acgt"), Seq("acgt")],
        )
        motif = motifs.Motif("ACGT", alignment)
```
wouldn't it be better if `motifs.Motif("ACGT", alignment)` raises an error, because it finds letters in the alignment that are missing from the alphabet?

**Comment by ernestprovo23** [comment]
good catch @mdehoon — re-reading, you're right that the construction is where this should be caught. an alphabet that doesn't cover the alignment letters is semantically broken input, not a runtime edge case.

just pushed a follow-up commit that adds an upfront check in `Motif.__init__`: if the alignment contains letters not in the alphabet, it raises `ValueError` with a message pointing at the case-mismatch case @zreitz hit (also catches partial mismatches like alignment with 'N' against an 'ACGT' alphabet).

one thing i'd like to push back on though — i think keeping the `total == 0` guard + `consensus_letter` initialization in `calculate_consensus` is still worth it as belt-and-suspenders, for a few reasons:

1. the all-zero column state is reachable through paths other than `Motif(alphabet, alignment)`: notably `Motif(alphabet, counts=...)` where the caller supplies the count matrix directly and the alignment validation never runs. trimming/filtering/manual construction can all land there too. an `UnboundLocalError` in any of those paths reads as a code smell to me.

2. this lines up with the layered-validation pattern in scikit-learn (validate in `fit`, re-check in `predict`) and pandas (`Series.idxmax()` on all-NaN raises a clear `ValueError` rather than letting a deep traceback bubble out). returning the undefined character for an all-zero column is the bio analog of returning NaN for a degenerate numerical case.

3. the relevant 2012 biopython-dev thread (bug #2597) that proposed enforcing alphabet letters was explicit that this catches "letters in an unexpected case" — i've implemented that — but the proposal didn't argue against also hardening downstream consumers. closing the construction-time hole and keeping the runtime guard feel complementary rather than redundant.

i added a test that exercises the defensive consensus guard via the counts= construction path (which bypasses alignment validation), so the test surface explicitly covers both layers. happy to drop the runtime guard if you'd rather keep the surface minimal, but the layered argument felt principled enough to defend.

ai disclosure: i used ai-assisted tooling for portions of the literature research (scikit-bio / scikit-learn / pandas precedents) and to help draft the test scaffolding. all design decisions, the validation logic, and the final review are mine.

**Comment by ernestprovo23** [comment]
hey @mdehoon, following up on the last comment... does the layered-validation argument hold up, or do you want me to trim the calculate_consensus guard and keep just the constructor check? happy to go either way, just waiting on your call.

File: Tests/test_motifs.py [hunk 1]
```diff
@@ -235,6 +235,81 @@ def test_reverse_complement(self):
...
+        instead of silently building an all-zero motif.
+        """
+        from Bio.Align import Alignment
+
+        alignment = Alignment(
+            [Seq("acgt"), Seq("acgt")],
+        )
+        with self.assertRaises(ValueError) as cm:
+            motifs.Motif("ACGT", alignment)
+        msg = str(cm.exception)
```
Comment (mdehoon) [inline Tests/test_motifs.py]: Can you check the exact `msg`? The `assertIn` may work, but it is more difficult to read, and may not notice other mistakes in the message.

File: Tests/test_motifs.py [hunk 2]
```diff
@@ -235,6 +235,81 @@ def test_reverse_complement(self):
...
+        rather than silently producing zero counts for the wrong letter.
+        """
+        from Bio.Align import Alignment
+
+        alignment = Alignment(
+            [Seq("ACGTN"), Seq("ACGTN")],
+        )
+        with self.assertRaises(ValueError) as cm:
+            motifs.Motif("ACGT", alignment)
+        self.assertIn("'N'", str(cm.exception))
```
Comment (mdehoon) [inline Tests/test_motifs.py]: Same comment here.

File: Tests/test_motifs.py [hunk 3]
```diff
@@ -235,6 +235,81 @@ def test_reverse_complement(self):
...
+        manual trimming or filtering), so calculate_consensus must
+        defensively return the undefined character instead of raising
+        UnboundLocalError.
+        """
+        zero_counts = {
+            "A": [0, 0, 0, 0],
+            "C": [0, 0, 0, 0],
+            "G": [0, 0, 0, 0],
+            "T": [0, 0, 0, 0],
+        }
```
Comment (mdehoon) [inline Tests/test_motifs.py:303]: To avoid confusion, can you choose some other sequence length instead of 4? Either 3 or 5, or whatever, as long as it doesn't happen to match the length of the alphabet.

File: Tests/test_motifs.py [hunk 4]
```diff
@@ -235,6 +235,81 @@ def test_reverse_complement(self):
...
+    def test_zero_counts_with_identity(self):
+        """Test that all-zero counts with identity parameter also works."""
+        zero_counts = {
+            "A": [0, 0, 0, 0],
+            "C": [0, 0, 0, 0],
+            "G": [0, 0, 0, 0],
+            "T": [0, 0, 0, 0],
+        }
+        motif = motifs.Motif("ACGT", counts=zero_counts)
+        result = motif.counts.calculate_consensus(identity=0.5)
```
Comment (mdehoon) [inline Tests/test_motifs.py]: You may as well merge this test with the previous one. A test typically has multiple assert statements.

File: Tests/test_motifs.py [hunk 5]
```diff
@@ -235,6 +235,81 @@ def test_reverse_complement(self):
...
+        self.assertEqual(result, "NNNN")
+
+    def test_normal_consensus_unchanged(self):
+        """Test that normal consensus behavior is not affected by the fix."""
```
Comment (mdehoon) [inline Tests/test_motifs.py]: This test should come first, as it tests the normal consensus behavior. Also, the comment can be just "Test normal consensus behavior" or something similar; a few months from now we won't know what we meant by "the fix".

**Comment by mdehoon** [comment]
I am OK with the current PR, except for a few minor issues (see above).

**Comment by ernestprovo23** [comment]
thanks michiel.

i did add the construction-time raise you flagged (commits 6fb9771 / 9f34a9, tolerating `-` and `.` gaps). before i push anything else though, one thing i want your read on, because it goes a bit past the original bug:

the check currently raises on *any* letter outside the alphabet, so against an ACGT alphabet it now also rejects alignments containing N or iupac ambiguity codes, which previously just built fine (those letters were ignored). on the current branch:

```python
aln = Alignment([Seq("ACGN"), Seq("ACGT")])
motifs.Motif("ACGT", aln)   # ValueError: ... not in alphabet 'ACGT': ['N']
```

N in a dna alignment is common enough that i'd rather not break it silently. two options:

a) narrow the check to case-mismatch only (extraneous letters that are case-variants of alphabet letters) ... keeps the #5000 catch, leaves N/iupac working.

b) drop the validation and keep just the minimal calculate_consensus guard, so the pr stays scoped to the unboundlocalerror.

fine with either, your call. and if the "minor issues" you meant are something else, point me at them and i'll roll them into the same push.

**Comment by mdehoon** [comment]
@ernestprovo23  The minor issues are just the comments on the specific lines in the code, shown above (starting from "Can you check the exact msg").

**Comment by ernestprovo23** [comment]
hey @mdehoon, i don't think those line comments made it through on my end... github keeps inline comments in a pending review until you hit "submit review", and right now nothing's showing on the diff for me (no line notes at all). mind submitting the review or pasting the notes here? happy to turn it around quick once i can see them.

my guess is the "exact msg" one is about the ValueError text in motifs/__init__.py. if so i'm glad to tighten the wording, just lmk the direction you'd prefer.

**Comment by ernestprovo23** [comment]
thanks michiel, all five addressed: both error-message tests now use exact assertEqual checks, dropped the column count off 4 so it doesn't shadow the alphabet length, merged the identity case into the zero-counts test, and moved the normal-consensus test ahead of the edge cases with a plainer comment. green locally, lmk if anything else.

**Comment by peterjc** [comment]
> ai disclosure: i used ai-assisted tooling for portions of the literature research (scikit-bio / scikit-learn / pandas precedents) and to help draft the test scaffolding. all design decisions, the validation logic, and the final review are mine.

If we decide to block AI generated code (see thread https://lists.open-bio.org/pipermail/biopython/2026-April/017113.html which is still on-going), are you willing to remove or re-write the new tests?

**Comment by ernestprovo23** [comment]
yes, happy to remove or re-write the tests by hand if the policy lands that way. just lmk and i'll turn it around.

**Comment by peterjc** [comment]
Thank you!

**Review by mdehoon — APPROVED** [review]
Code looks good to me, thanks!
The CircleCI failure seems unrelated.

**Comment by peterjc** [comment]
Agreed, the coverage upload from CircleCI is broken right now.

I need to draft the no-AI policy I have in mind as a PR for discussion... if you wanted to redo the tests by hand now, we could avoid the AI issue and merge this.
